node/bindnode/repr.go

```diff
@@ -0,0 +1,814 @@
+	// TODO: why doesn't stg just allow us to "get" by the key string?
+	typ := schema.SpawnBool(schema.TypeName(key))
+	mappedKey := stg.GetDiscriminant(typ)
+	return mappedKey
+}
+
+func inboundMappedType(typ *schema.TypeUnion, stg schema.UnionRepresentation_Keyed, key string) string {
+	// TODO: can't do a "reverse" lookup... needs better API probably.
+	for _, member := range typ.Members() {
+		mappedKey := stg.GetDiscriminant(member)
+		if key == mappedKey {
+			// println(key, "rev-mapped to", field.Name())
+			return member.Name().String()
+		}
+	}
+	// println(key, "had no mapping")
+	return key // fallback to the same key
+}
+
+func (w *_nodeRepr) asKinded(stg schema.UnionRepresentation_Kinded, kind ipld.Kind) *_nodeRepr {
+	name := stg.GetMember(kind)
+	members := w.schemaType.(*schema.TypeUnion).Members()
+	for _, member := range members {
+		if member.Name() != name {
+			continue
+		}
+		w2 := *w
+		w2.val = w.val.FieldByName("Value").Elem()
+		w2.schemaType = member
+		return &w2
+	}
+	return nil
+}
+
+func (w *_nodeRepr) LookupByString(key string) (ipld.Node, error) {
+	if stg, ok := reprStrategy(w.schemaType).(schema.UnionRepresentation_Kinded); ok {
+		w = w.asKinded(stg, ipld.Kind_Map)
+	}
+	switch stg := reprStrategy(w.schemaType).(type) {
+	case schema.StructRepresentation_Map:
+		revKey := inboundMappedKey(w.schemaType.(*schema.TypeStruct), stg, key)
+		v, err := (*_node)(w).LookupByString(revKey)
+		if err != nil {
+			return nil, err
+		}
+		return reprNode(v), nil
+	case schema.UnionRepresentation_Keyed:
+		revKey := inboundMappedType(w.schemaType.(*schema.TypeUnion), stg, key)
+		v, err := (*_node)(w).LookupByString(revKey)
+		if err != nil {
+			return nil, err
+		}
+		return reprNode(v), nil
+	case nil:
+		v, err := (*_node)(w).LookupByString(key)
+		if err != nil {
+			return nil, err
+		}
+		return reprNode(v), nil
+	default:
+		panic(fmt.Sprintf("TODO: %T", stg))
+	}
+}
+
+func (w *_nodeRepr) LookupByNode(key ipld.Node) (ipld.Node, error) {
+	return nil, ipld.ErrWrongKind{
+		TypeName:   w.schemaType.Name().String(),
+		MethodName: "LookupByNode", AppropriateKind: ipld.KindSet_JustList, ActualKind: ipld.Kind_Map,
+	}
+}
+
+func (w *_nodeRepr) LookupByIndex(idx int64) (ipld.Node, error) {
+	switch stg := reprStrategy(w.schemaType).(type) {
+	case schema.StructRepresentation_Tuple:
+		fields := w.schemaType.(*schema.TypeStruct).Fields()
+		field := fields[idx]
+		v, err := (*_node)(w).LookupByString(field.Name())
+		if err != nil {
+			return nil, err
+		}
+		return reprNode(v), nil
+	case nil:
+		v, err := (*_node)(w).LookupByIndex(idx)
+		if err != nil {
+			return nil, err
+		}
+		return reprNode(v), nil
+	default:
+		panic(fmt.Sprintf("TODO: %T", stg))
+	}
+}
+
+func (w *_nodeRepr) LookupBySegment(seg ipld.PathSegment) (ipld.Node, error) {
+	return nil, ipld.ErrWrongKind{
+		TypeName:   w.schemaType.Name().String(),
+		MethodName: "LookupBySegment", AppropriateKind: ipld.KindSet_JustList, ActualKind: ipld.Kind_Map,
```

Comment:
Should be `AppropriateKind: ipld.KindSet_Recursive` among other things.  Probably this whole method is actually a todo right now?

Reply:
Yeah, a lot of these are copy pasted and TODOs. I'll make it explicit.